### PR TITLE
[Email Skill]Adaptive Card update with Graph API

### LINKED ISSUE #1028: [Email Skill]Adaptive Card update with Graph API
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by DingmaomaoBJTU**
## Description

Close #1028: [Email Skill]Adaptive Card update with Graph API.

### Conversational Experience
Show email list card:
image

Show email detail card:
image

Search email by sender card:
image

Search email by contact card:
image

Message content confirmation:
image

Email confirmation:
image

PR include the following:
1. Add msgraph user photo retrieve API support to get user photo.
2. Apply all email adaptive cards to existing flows and bind related data, including:
1) email overview card
2) email overview card with condition
3) email detail card
4) confirm message card
5) confirm email card

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the appropriate unit tests
- [x] I have tested any new/updated dialogs using Speech in the emulator to ensure the [Speak](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-text-to-speech?view=azure-bot-service-3.0) property is set to enable a high quality speech-first experience and the appropriate [InputHints](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-add-input-hints?view=azure-bot-service-3.0) are set correctly. 
- [x] I have updated related documentation

- [x] A duplicate issue is filed to track future  work.

- [ ] All languages have been updated
- [ ] You have tested deployment with your new models

### CODE REVIEW

File: lib/microsoft.bot.builder.solutions/Responses/ResponseManager.cs [hunk 1]
```diff
@@ -152,6 +152,46 @@ public Activity GetCardResponse(Card card)
             return MessageFactory.Carousel(attachments, response.Text, response.Speak, response.InputHint) as Activity;
         }
 
+        public Activity GetCardResponse(
+            string templateId,
+            Card card,
+            StringDictionary tokens = null,
+            string containerName = null,
+            IEnumerable<Card> containerItems = null)
```
Comment (DingmaomaoBJTU) [inline lib/microsoft.bot.builder.solutions/Responses/ResponseManager.cs]: Will remove this after the new solution package is published.

**Comment by ryanisgrig** [comment]
I pushed a small change to add spacing on the Messages/High Priority Messages container so the background could fit for the overview by condition cards.

image
 #Closed

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailContentPreview.json [hunk 2]
```diff
@@ -1,20 +1,65 @@
...
-          "maxLines": 5
+          "wrap": true
         }
       ]
     }
-  ]
+  ],
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "version": "1.0",
+  "speak": ""
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailContentPreview.json]: Removed the speak here? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json [hunk 3]
```diff
@@ -1,72 +1,66 @@
 ﻿{
-  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
   "type": "AdaptiveCard",
-  "version": "1.0",
-  "speak": "{Speak}",
   "id": "EmailDetailCard",
+  "height": "stretch",
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json]: We can remove all instances of `"height":"stretch"` at the root-level of cards, I don't see this doing anything in this version. #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json [hunk 4]
```diff
@@ -77,5 +71,8 @@
       "title": "View",
       "url": "{EmailLink}"
     }
-  ]
+  ],
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "version": "1.0",
+  "speak": ""
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json]: Speak is missing? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewByCondition.json [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                          "weight": "Bolder",
+                          "color": "Light",
+                          "text": "{MessagesDiscription}"
+                        },
+                        {
+                          "type": "TextBlock",
+                          "size": "Medium",
+                          "weight": "Bolder",
+                          "color": "Light",
+                          "text": "{ImportantMessagesDiscription}"
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewByCondition.json]: Typo, rename to `ImportantMessagesDescription` #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewByCondition.json [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                      "type": "Column",
+                      "height": "stretch",
+                      "items": [
+                        {
+                          "type": "TextBlock",
+                          "height": "stretch",
+                          "size": "Medium",
+                          "weight": "Bolder",
+                          "color": "Light",
+                          "text": "{MessagesDiscription}"
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewByCondition.json]: Typo, rename to `MessagesDescription` #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+        public string TotalMessageNumber { get; set; }
+
+        public string HighPriorityMessagesNumber { get; set; }
+
+        public string Now { get; set; }
+
+        public string MailSourceType { get; set; }
+
+        public string MessagesDiscription { get; } = EmailCommonStrings.Messages;
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs]: `MessagesDescription` #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+        public string ImportantMessagesDiscription { get; } = EmailCommonStrings.ImportantMessages;
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs]: `ImportantMessagesDescription` #Resolved

**Review by ryanisgrig — CHANGES_REQUESTED** [review]

**Comment by DingmaomaoBJTU** [comment]
Thank you so much!!!!

---
In reply to: [477254976](https://github.com/Microsoft/AI/pull/1029#issuecomment-477254976) [](ancestors = 477254976)

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json (hunk 3 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json]: Sure~~Thanks!

---
In reply to: [269719790](https://github.com/Microsoft/AI/pull/1029#discussion_r269719790) [](ancestors = 269719790)

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json (hunk 4 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json]: Will remove that~

---
In reply to: [269719944](https://github.com/Microsoft/AI/pull/1029#discussion_r269719944) [](ancestors = 269719944)

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewByCondition.json (hunk 6 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewByCondition.json]: Thanks!

---
In reply to: [269720147](https://github.com/Microsoft/AI/pull/1029#discussion_r269720147) [](ancestors = 269720147)

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json [hunk 9]
```diff
@@ -75,7 +30,10 @@
     {
       "type": "Action.OpenUrl",
       "title": "View",
-      "url": "{EmailLink}"
+      "url": "https://outlook.office365.com/owa/?ItemID=AAMkAGEwMDg4MTVlLTVjYzUtNGRmNS1hYTJkLTI3NjNmM2Q3MDA3ZABGAAAAAABw%2FWbciOIvSZejmvtHK98CBwCa0ISfUgx1RYIZgSZZYh9LAAAAAAEMAACa0ISfUgx1RYIZgSZZYh9LAAAAtDQrAAA%3D&exvsurl=1&viewmodel=ReadMessageItem"
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json]: this url shouldn't be checked in? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Util/Util.cs [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    public class Util
+    {
+        public static bool IsEmail(string email)
+        {
+            if (string.IsNullOrWhiteSpace(email))
+            {
+                return false;
+            }
+
+            return Regex.IsMatch(email, @"^\w+([-+.]\w+)*@\w+([-.]\w+)*\.\w+([-.]\w+)*$");
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Util/Util.cs:19]: provide a comment explaining what the regex is/is-not looking for #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json (hunk 9 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailDetailCard.json]: Good catch, thanks!

---
In reply to: [270674200](https://github.com/Microsoft/AI/pull/1029#discussion_r270674200) [](ancestors = 270674200)

File: .gitignore [hunk 11]
```diff
@@ -345,3 +345,5 @@ ASALocalRun/
 /solutions/**/.deployment
 /solutions/**/publish
 /solutions/**/publish.cmd
+/templates/Virtual-Assistant-Template/VirtualAssistantTemplate/VirtualAssistantTemplate/skills.development.json
```
Comment (ryanisgrig) [inline .gitignore]: Will these be temporary based on current master branch or always in the `.gitignore` file? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/FindContact/FindContactDialog.cs [hunk 12]
```diff
@@ -144,7 +145,7 @@ public async Task<DialogTurnResult> AfterUpdateUserName(WaterfallStepContext sc,
                     return await sc.EndDialogAsync();
                 }
 
-                if (IsEmail(userInput))
+                if (Util.Util.IsEmail(userInput))
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/FindContact/FindContactDialog.cs:148]: Typo? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailContentPreview.json [hunk 13]
```diff
@@ -1,20 +1,76 @@
 ﻿{
-  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
   "type": "AdaptiveCard",
-  "version": "1.0",
-  "speak": "{Speak}",
-  "id": "EmailContentPreview",
+  "id": "EmailCard",
+  "height": "stretch",
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailContentPreview.json]: We can remove `"height":"stretch"` from the root level of all cards, I think this is a copy/paste error and the adaptive card designer doesn't care if it is there. #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
+﻿using EmailSkill.Dialogs.Shared.Resources.Strings;
+using Microsoft.Bot.Builder.Solutions.Responses;
+
+namespace EmailSkill.Dialogs.Shared.Resources.Cards
+{
+    public class EmailOverviewData : ICardData
+    {
+        public string Description { get; set; }
+
+        public string AvatorIcon { get; set; }
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs:10]: Rename to `AvatarIcon` #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+        public string TotalMessageNumber { get; set; }
```
Comment (ryanisgrig) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs:12]: Rename to `TotalMessagesNumber` #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/FindContact/FindContactDialog.cs (hunk 12 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/FindContact/FindContactDialog.cs:148]: No, this is a static function from Util namespace of Util class~

---
In reply to: [271027099](https://github.com/Microsoft/AI/pull/1029#discussion_r271027099) [](ancestors = 271027099)

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs (hunk 14 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Cards/EmailOverviewData.cs:10]: Thanks! Sorry for the typo ^_^~~

---
In reply to: [271053774](https://github.com/Microsoft/AI/pull/1029#discussion_r271053774) [](ancestors = 271053774)

File: .gitignore (hunk 11 above)
Comment (DingmaomaoBJTU) [reply .gitignore]: Confirmed with Lauren, these are for local changes. Will remove, thanks!

---
In reply to: [271026765](https://github.com/Microsoft/AI/pull/1029#discussion_r271026765) [](ancestors = 271026765)

**Review by ryanisgrig — APPROVED** [review]

**Pull request merged**